evdns.c

```diff
@@ -1738,18 +1959,173 @@ evdns_add_server_port_with_base(struct event_base *base, evutil_socket_t socket,
 	port->user_data = user_data;
 	port->pending_replies = NULL;
 	port->event_base = base;
-
-	event_assign(&port->event, port->event_base,
-				 port->socket, EV_READ | EV_PERSIST,
-				 server_port_ready_callback, port);
-	if (event_add(&port->event, NULL) < 0) {
-		mm_free(port);
-		return NULL;
+	port->max_client_connections = MAX_CLIENT_CONNECTIONS;
+	port->client_connections_count = 0;
+
+	if (sock_type == SOCK_DGRAM) {
+		/* UDP DNS server */
+		event_assign(&port->event, port->event_base,
+					 port->socket, EV_READ | EV_PERSIST,
+					 server_port_ready_callback, port);
+		if (event_add(&port->event, NULL) < 0) {
+			mm_free(port);
+			return NULL;
+		}
+	} else {
+		/* TCP DNS server */
+		port->socket = -1;
+		port->listener = evconnlistener_new(base, incoming_conn_cb, port,
+							LEV_OPT_CLOSE_ON_FREE | LEV_OPT_REUSEABLE, -1, socket);
+		if (!port->listener) {
+			mm_free(port);
+			return NULL;
+		}
 	}
+	LIST_INIT(&port->client_connections);
+
 	EVTHREAD_ALLOC_LOCK(port->lock, EVTHREAD_LOCKTYPE_RECURSIVE);
 	return port;
 }
 
+static void
+server_tcp_event_cb(struct bufferevent *bev, short events, void *ctx);
+
+static int
+tcp_read_message(struct tcp_connection *conn, u8 **msg, int *msg_len)
+{
+	struct bufferevent *bev = conn->bev;
+	struct evbuffer *input = bufferevent_get_input(bev);
+	u8 *packet = NULL;
+	int r = 0;
+
+	EVUTIL_ASSERT(conn);
+	EVUTIL_ASSERT(conn->state == TS_CONNECTED);
+
+	/* reading new packet size */
+	if (!conn->awaiting_packet_size) {
+		if (evbuffer_get_length(input) < 2)
+			goto awaiting_next;
+
+		bufferevent_read(bev, (void*)&conn->awaiting_packet_size,
+			sizeof(conn->awaiting_packet_size));
+		conn->awaiting_packet_size = ntohs(conn->awaiting_packet_size);
+		if (conn->awaiting_packet_size <= 0)
+			goto fail;
+	}
+
+	/* reading new packet content */
+	if (evbuffer_get_length(input) < conn->awaiting_packet_size)
+		goto awaiting_next;
+
+	packet = mm_malloc(conn->awaiting_packet_size);
+	if (!packet)
+		goto fail;
+
+	r = (int)bufferevent_read(bev, (void*)packet, conn->awaiting_packet_size);
+	if (r != conn->awaiting_packet_size) {
+		mm_free (packet);
+		packet = NULL;
+		goto fail;
+	}
+
+	*msg = packet;
+	*msg_len = r;
+awaiting_next:
+	return 0;
+fail:
+	return 1;
+}
+
+static void
+server_tcp_read_packet_cb(struct bufferevent *bev, void *ctx)
+{
+	u8 *msg = NULL;
+	int msg_len = 0;
+	struct client_tcp_connection *client = (struct client_tcp_connection *)ctx;
+	struct evdns_server_port *port = client->port;
+	struct tcp_connection *conn = &client->connection;
+	EVUTIL_ASSERT(port && bev);
+	EVDNS_LOCK(port);
+
+	while (1) {
+		if (tcp_read_message(conn, &msg, &msg_len)) {
+			log(EVDNS_LOG_MSG, "Closing client connection %p due to error", bev);
+			evdns_remove_tcp_client(port, client);
```

Comment:
this will not notify the user and fail the nameserver?

Reply:
We were not able to read the incoming message here. I don't understand how we can notify the user here. 
Could you please add some details? I think I don't quite understand your idea here.

Reply:
I think the desire is to notify some user callback in  a way that indicates error, so that it can clean-up and maybe retry etc.

Reply:
Hi, thanks for clarification.
This code belongs to server. I think call servers callbacks here is strange since DNS client might have sent some garbage or some other error might have happened. So the right choice is simple destroy the connection.
To send some reasonable reply to the DNS client is also hard here since we haven't read the message and don't know `id` of the reply. I'll read RFCs maybe there are some special type of messages for TCP connections that we can use to notify the client in case we were not able to read its message. But I doubt I'll found anything.

Reply:
Ah, yeah that makes sense, I didn't looks closely at all. But I can suggest a reason that makes more sense: so the server application can count or log errors, including remote-ip-addr.

Reply:
> I'll read RFCs maybe there are some special type of messages for TCP connections that we can use to notify the client in case we were not able to read its message. But I doubt I'll found anything.

Ok, didn't know that, and FWIW checked dnsmasq/systemd-resolv/unbound all of them do not reply anything.